packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js

```diff
@@ -4203,4 +4203,60 @@ describe('ReactSuspenseWithNoopRenderer', () => {
     });
     expect(Scheduler).toHaveYielded(['Unmount Child']);
   });
+
+  it('should resume the suspending component when work loop yields immediately after catching wakeable and when wakeable pings before resuming the work loop', async () => {
+    let cachedFoo;
+    let cachedBar;
+    const resolveCbs = new Set();
+    function useFoo() {
+      if (!cachedFoo) {
+        throw new Promise(resolve =>
+          setTimeout(() => resolve((cachedFoo = 'foo')), 50),
+        );
+      }
+      return cachedFoo;
+    }
+    function useBar() {
+      if (!cachedBar) {
+        const p = {
+          then(cb) {
+            resolveCbs.add(cb);
+            // uncomment to make this test pass
+            // if (cachedBar) {
+            //   p.resolve();
+            // }
+          },
+          resolve() {
+            resolveCbs.forEach(cb => cb());
+          },
+        };
+        Scheduler.unstable_forceYield();
+        cachedBar = 'bar';
+        p.resolve();
```

Comment:
This is slightly different from @bvaughn's repro but it leads to the same behavior and ultimately it doesn't matter. Here I resolve this `p` **synchronously** so I lead to a synchronous ping - In Brian's case the ping was asynchronous, it managed to get in between the yield and work loop resuming its job.

Reply:
Actually... this doesn't match @bvaughn's case 1 to 1. The problem is that here the wakeable is resolved before even throwing, so React doesn't even have a chance to attach to its `then`. With this test case we don't get to "ping" React at all - but the real issue was that React got "pinged", one of the React-attached then callbacks was resolved, but the other one wasn't.